Libraries/LibIPC/Connection.h

```diff
@@ -81,38 +94,55 @@ class Connection : public ConnectionBase {
     {
     }
 
-    template<typename MessageType>
-    OwnPtr<MessageType> wait_for_specific_message()
+    ~Connection() override = default;
+
+    template<typename RequestType, typename... Args>
+    NonnullOwnPtr<typename RequestType::ResponseType> send_sync(Args&&... args)
     {
-        return wait_for_specific_endpoint_message<MessageType, LocalEndpoint>();
+        auto result = send<RequestType, Args...>(forward<Args>(args)...)->await();
+        VERIFY(!result.is_error());
+        auto ok = result.release_value();
+        VERIFY(ok);
+        return ok.release_nonnull();
     }
 
     template<typename RequestType, typename... Args>
-    NonnullOwnPtr<typename RequestType::ResponseType> send_sync(Args&&... args)
+    NonnullRefPtr<Core::Promise<OwnPtr<typename RequestType::ResponseType>>> send(Args&&... args)
     {
-        MUST(post_message(RequestType(forward<Args>(args)...)));
-        auto response = wait_for_specific_endpoint_message<typename RequestType::ResponseType, PeerEndpoint>();
-        VERIFY(response);
-        return response.release_nonnull();
+        using Promise = Core::Promise<OwnPtr<typename RequestType::ResponseType>>;
+        auto promise = Promise::construct();
+        auto const msg = RequestType { LocalEndpoint::next_ipc_request_id(), forward<Args>(args)... };
+        MUST(post_message(msg));
+        m_resolvers.empend(msg.ipc_request_id(), [promise, msg_id = msg.message_id()](OwnPtr<IPC::Message> msg) {
+            ASSERT(msg->message_id() == msg_id + 1 && msg->endpoint_magic() == PeerEndpoint::static_magic());
```

Comment:
We tend to reserve `ASSERT` for performance-critical code. I don't think this assertion (and a bunch below) are going to cause any perf issues, so to be safer, we'd stick with VERIFY here.

We would also break each of these into their own assertions, so if it breaks, we know which one:
```c++
VERIFY(msg->message_id() == msg_id + 1);
VERIFY(msg->endpoint_magic() == PeerEndpoint::static_magic());
```